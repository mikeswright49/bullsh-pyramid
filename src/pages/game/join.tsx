import React from 'react';

export default function JoinGame(): JSX.Element {
    return <h1>Join Game</h1>;
}
