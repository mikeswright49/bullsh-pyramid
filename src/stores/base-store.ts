import firebase from 'firebase';
import { getFirebaseConfig } from 'config/firebase-config';

export abstract class BaseStore {
    public static database: firebase.database.Database;

    public static init() {
        if (!firebase.apps.length) {
            firebase.initializeApp(getFirebaseConfig());
            firebase.auth().signInAnonymously();
        }
        if (!BaseStore.database) {
            BaseStore.database = firebase.database();
        }
    }
}
