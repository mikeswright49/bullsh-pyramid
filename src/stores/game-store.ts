import * as firebase from 'firebase';
import { shortId } from '../utilities/shortid';
import { GameStage } from '../enums/game-stage';
import { GameState } from 'types/game-state';
import { getFirebaseConfig } from '../../config/firebase-config';
import { Player } from 'types/player';
import { Card } from 'types/card';

export class GameStore {
    private static database: firebase.database.Database;
    private static subscribers = [];

    public static async createGame(playerCount: number, tierCount: number): Promise<string> {
        GameStore.init();
        const ID_LENGTH = 4;
        const gameId = shortId(ID_LENGTH);

        const gameState = {
            id: gameId,
            gameStage: GameStage.Initiation,
            activeIndex: 0,
            activeRow: 0,
            playerCount,
            tierCount,
        } as GameState;

        try {
            await GameStore.database.ref(`games/${gameId}`).set(gameState);
            return gameId;
        } catch (error) {
            console.error(error);
        }
    }

    public static subscribeToGame(gameId: string, updateCallback: (snapshot: GameState) => void) {
        GameStore.init();
        const game = GameStore.database.ref(`games/${gameId}`);
        GameStore.subscribers.push(
            game.on('value', (val) => {
                const value = val.val();
                updateCallback(value as GameState);
            })
        );
    }

    public static subscribeToPlayers(gameId: string, updateCallback: (snapshot: Player) => void) {
        GameStore.init();
        const game = GameStore.database.ref(`games/${gameId}/players`);
        const players = GameStore.database.ref(`players`);
        game.on('child_added', (child) => {
            const playerId = child.val();
            players.child(playerId).on('value', (val) => {
                const value = val.val();
                updateCallback(value as Player);
            });
        });
    }

    public static unsubscribeToGame() {
        GameStore.subscribers.forEach((sub) => sub.off('value'));
    }

    public static async updateGameStage(gameId: string, gameStage: GameStage) {
        await GameStore.database.ref(`games/${gameId}`).update({
            gameStage,
        });
    }

    public static async updateTiers(gameId: string, tiers: Card[][]) {
        await GameStore.database.ref(`games/${gameId}`).update({
            tiers,
        });
    }

    public static async setActiveCard(
        gameId: string,
        index: { activeRow: number; activeIndex: number }
    ) {
        await GameStore.database.ref(`games/${gameId}`).update(index);
    }

    public static async joinGame(gameId: string, playerId: string) {
        await GameStore.database.ref(`games/${gameId}/players`).push(playerId);
    }

    private static init() {
        if (!firebase.apps.length) {
            firebase.initializeApp(getFirebaseConfig());
        }
        if (!GameStore.database) {
            GameStore.database = firebase.database();
        }
    }
}
