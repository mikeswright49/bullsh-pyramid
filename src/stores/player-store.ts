import * as firebase from 'firebase';
import { getFirebaseConfig } from '../../config/firebase-config';
import { shortId } from '../utilities/shortid';
import { Card } from 'types/card';
import { Player } from 'types/player';

export class PlayerStore {
    private static DEFAULT_PLAYER_SCORE = 10000;
    private static database: firebase.database.Database;
    private static subscribers = [];

    public static subscribeToPlayer(playerId: string, updateCallback: (snapshot: Player) => void) {
        const player = PlayerStore.database.ref(`players/${playerId}`);
        PlayerStore.subscribers.push(
            player.on('value', (val: { val: () => Player }) => {
                const value = val.val();
                updateCallback(value as Player);
            })
        );
    }

    public static subscribeToHaters(playerId: string, updateCallback: (hater: Player) => void) {
        const player = PlayerStore.database.ref(`players/${playerId}/hatersmap`);
        const players = PlayerStore.database.ref(`players`);
        PlayerStore.subscribers.push(
            player.on('child_added', (child) => {
                const playerId = child.val();
                players.child(playerId).once('value', (val) => {
                    const value = val.val();
                    updateCallback(value as Player);
                });
            })
        );
    }

    public static async updatePlayer(player: Player) {
        await PlayerStore.database.ref(`players/${player.id}`).update(player);
    }

    public static unsubscribeToPlayer() {
        PlayerStore.subscribers.forEach((sub) => sub.off('value'));
    }

    public static async createPlayer(name: string): Promise<string> {
        const playerId = shortId();

        const player: Player = {
            id: playerId,
            name,
            hand: [],
            score: PlayerStore.DEFAULT_PLAYER_SCORE,
            isHost: false,
            declaration: null,
            hasVoted: false,
            haters: [],
            hatersmap: {},
        };

        try {
            await PlayerStore.database.ref(`players/${playerId}`).set(player);
            return playerId;
        } catch (error) {
            console.error(error);
        }
    }

    public static async setDeclaration(playerId: string, card: Card) {
        try {
            await PlayerStore.database.ref(`players/${playerId}`).update({
                declaration: card,
                hasVoted: true,
            });
        } catch (e) {
            console.error(e);
        }
    }

    public static async setHasVoted(playerId: string, hasVoted: boolean) {
        try {
            await PlayerStore.database.ref(`players/${playerId}`).update({
                hasVoted,
            });
        } catch (e) {
            console.error(e);
        }
    }

    public static async addHater(playerId: string, haterId: string) {
        try {
            await PlayerStore.database.ref(`players/${playerId}/hatersmap`).push(haterId);
        } catch (e) {
            console.error(e);
        }
    }

    public static init() {
        if (!firebase.apps.length) {
            firebase.initializeApp(getFirebaseConfig());
        }
        PlayerStore.database = firebase.database();
    }
}
