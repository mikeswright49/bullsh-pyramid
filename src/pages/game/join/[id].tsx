import React from 'react';
import { useRouter } from 'next/router';
import { Join } from 'src/routes/join';

export default function JoinGame(): JSX.Element {
    const router = useRouter();
    const gameId = router.query.id as string;
    return <Join gameId={gameId} />;
}
