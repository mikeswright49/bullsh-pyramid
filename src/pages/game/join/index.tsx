import React, { useRef } from 'react';
import { Layout } from '../../../components/layout/layout';
import { PlayerStore } from '../../../stores/player-store';
import { GameStore } from '../../../stores/game-store';

export default function JoinGame(): JSX.Element {
    const gameIdRef = useRef<HTMLInputElement>();
    const playerNameRef = useRef<HTMLInputElement>();

    const createPlayerAndJoin = async (): Promise<void> => {
        const gameId = gameIdRef.current.value;
        const playerName = playerNameRef.current.value;
        const playerId = await PlayerStore.createPlayer(playerName);
        await GameStore.joinGame(gameId, playerId);
        window.location.assign(`/game/${gameId}/player/${playerId}`);
    };

    return (
        <Layout>
            <>
                <h1>Join Game</h1>
                <span>Enter your info and we can get this party started</span>
                <div className="stack-y-2">
                    <label>
                        Game id:
                        <input ref={gameIdRef} type="text" />
                    </label>
                </div>
                <div className="stack-y-2">
                    <label>
                        Player name:
                        <input ref={playerNameRef} type="text" />
                    </label>
                </div>
                <div>
                    <button
                        className="pure-button pure-button-primary"
                        onClick={createPlayerAndJoin}
                    >
                        Join the game
                    </button>
                </div>
            </>
        </Layout>
    );
}
