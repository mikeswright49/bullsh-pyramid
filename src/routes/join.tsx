import React, { useRef } from 'react';
import { GameStore } from 'src/stores/game-store';
import { PlayerStore } from 'src/stores/player-store';
import { Layout } from 'src/components/layout/layout';

export function Join({ gameId }: { gameId?: string }) {
    const gameIdRef = useRef<HTMLInputElement>({ value: gameId } as HTMLInputElement);
    const playerNameRef = useRef<HTMLInputElement>();

    const createPlayerAndJoin = async (): Promise<void> => {
        GameStore.init();

        const gameToJoin = gameId || gameIdRef.current.value;
        const playerName = playerNameRef.current.value;
        const playerId = await PlayerStore.createPlayer(playerName);
        await GameStore.joinGame(gameToJoin, playerId);
        window.location.assign(`/game/${gameToJoin}/player/${playerId}`);
    };

    return (
        <Layout>
            <>
                <h1>Join Game</h1>
                <span>Enter your info and we can get this party started</span>
                <div className="stack-y-2">
                    <label>
                        <span className="stack-x-1">Game id:</span>
                        {!gameId && <input ref={gameIdRef} type="text" />}
                        {!!gameId && <strong>{gameId}</strong>}
                    </label>
                </div>
                <div className="stack-y-2">
                    <label>
                        <span className="stack-x-1">Player name:</span>
                        <input ref={playerNameRef} type="text" />
                    </label>
                </div>
                <div>
                    <button
                        className="pure-button pure-button-primary"
                        onClick={createPlayerAndJoin}
                    >
                        Join the game
                    </button>
                </div>
            </>
        </Layout>
    );
}
