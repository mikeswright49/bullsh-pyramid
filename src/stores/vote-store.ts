import { BaseStore } from './base-store';
import { shortId } from 'src/utilities/shortid';
import { Vote } from 'types/vote';
import { Card } from 'types/card';
import { PlayerStore } from './player-store';
import { Player } from 'types/player';

export class VoteStore extends BaseStore {
    public static async createVote({
        amount,
        playerId,
        targetId,
        card,
    }: {
        amount: number;
        playerId: string;
        targetId: string;
        card: Card;
    }): Promise<string> {
        const voteId = shortId();
        try {
            await VoteStore.database.ref(`votes/${voteId}`).set({ amount, card });
            await VoteStore.database.ref(`votes/${voteId}/playerref`).push(playerId);
            await VoteStore.database.ref(`votes/${voteId}/targetref`).push(targetId);
            return voteId;
        } catch (error) {
            console.error(error);
        }
    }

    public static async addVote(gameId: string, voteId: string) {
        try {
            await VoteStore.database.ref(`/games/${gameId}/votes`).push(voteId);
        } catch (error) {
            console.error(error);
        }
    }

    public static async getTarget(voteId: string): Promise<Player> {
        const ref = VoteStore.database.ref(`votes/${voteId}/targetref`);
        return VoteStore.getPlayerRef(ref);
    }

    public static async getPlayer(voteId: string): Promise<Player> {
        const ref = VoteStore.database.ref(`votes/${voteId}/playerref`);
        return VoteStore.getPlayerRef(ref);
    }

    public static async removeVotes(gameId: string) {
        try {
            const votes = await VoteStore.database.ref(`/games/${gameId}/votes`).once('value');
            await Promise.all(
                votes.val().map((childId: string) => {
                    return votes.child(childId).ref.remove();
                })
            );
        } catch (error) {
            console.error(error);
        }
    }

    public static subscribeToVotes(gameId: string, callback: (vote: Vote) => void) {
        if (!gameId) {
            return;
        }
        const game = VoteStore.database.ref(`games/${gameId}/votes`);
        const votes = VoteStore.database.ref(`votes`);
        game.on('child_added', (child) => {
            const voteId = child.val();
            votes.child(voteId).on('value', async (val) => {
                const value: Vote = val.val();
                value.id = voteId;
                value.player = await VoteStore.getPlayer(voteId);
                value.target = await VoteStore.getTarget(voteId);
                callback(value as Vote);
            });
        });
    }

    private static async getPlayerRef(ref: firebase.database.Reference): Promise<Player> {
        const playerId: string = (await ref.once('child_added')).val();
        return PlayerStore.getPlayer(playerId);
    }
}
