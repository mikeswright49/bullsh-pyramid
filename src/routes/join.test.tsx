import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';
import { Join } from './join';
import { PlayerStore } from 'src/stores/player-store';
import { GameStore } from 'src/stores/game-store';

jest.mock('src/stores/player-store');
jest.mock('src/stores/game-store');

describe('<Unit Test> Join Route', () => {
    beforeEach(() => {
        (PlayerStore.createPlayer as jest.Mock).mockReturnValue(Promise.resolve('a-player-id'));

        const oldWindowLocation = window.location;
        delete window.location;

        window.location = Object.defineProperties(
            {},
            {
                ...Object.getOwnPropertyDescriptors(oldWindowLocation),
                assign: {
                    configurable: true,
                    value: jest.fn(),
                },
            }
        );
    });

    it('should render the form with a game id input if no id is provided', () => {
        const { container } = render(<Join />);
        expect(container.querySelector('[data-testid="join-page"]')).toMatchSnapshot();
    });
    it('should render a strong gameId if provided', () => {
        const { container } = render(<Join gameId={'1234'} />);
        expect(container.querySelector('[data-testid="join-page"]')).toMatchSnapshot();
    });
    it('should be able to create the player and add them to the game on submit', async () => {
        const { container } = render(<Join gameId={'1234'} />);
        fireEvent.change(container.querySelector('[data-testid="player-name"]'), {
            target: { value: 'frank' },
        });
        await act(async () =>
            fireEvent.submit(container.querySelector('[data-testid="join-form"]'))
        );
        expect(PlayerStore.createPlayer).toHaveBeenCalledWith('frank', true);
        expect(GameStore.joinGame).toHaveBeenCalledWith('1234', 'a-player-id');
        expect(container.querySelector('[data-testid="join-page"]')).toMatchSnapshot();
        expect(window.location.assign).toHaveBeenCalledWith('/game/1234/player/a-player-id');
    });
});
