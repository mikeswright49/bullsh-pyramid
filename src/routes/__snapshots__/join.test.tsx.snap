// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Unit Test> Join Route should be able to create the player and add them to the game on submit 1`] = `
<div
  class="container"
  data-testid="join-page"
>
  <h1>
    Join Game
  </h1>
  <span>
    Enter your info and we can get this party started
  </span>
  <form
    data-testid="join-form"
  >
    <div
      class="stack-y-2"
    >
      <label>
        <span
          class="stack-x-1"
        >
          Game id:
        </span>
        <strong>
          1234
        </strong>
      </label>
    </div>
    <div
      class="stack-y-2"
    >
      <label>
        <span
          class="stack-x-1"
        >
          Player name:
        </span>
        <input
          data-testid="player-name"
          type="text"
        />
      </label>
    </div>
    <div>
      <button
        class="pure-button pure-button-primary"
        data-testid="join-submit"
      >
        Join the game
      </button>
    </div>
  </form>
</div>
`;

exports[`<Unit Test> Join Route should render a strong gameId if provided 1`] = `
<div
  class="container"
  data-testid="join-page"
>
  <h1>
    Join Game
  </h1>
  <span>
    Enter your info and we can get this party started
  </span>
  <form
    data-testid="join-form"
  >
    <div
      class="stack-y-2"
    >
      <label>
        <span
          class="stack-x-1"
        >
          Game id:
        </span>
        <strong>
          1234
        </strong>
      </label>
    </div>
    <div
      class="stack-y-2"
    >
      <label>
        <span
          class="stack-x-1"
        >
          Player name:
        </span>
        <input
          data-testid="player-name"
          type="text"
        />
      </label>
    </div>
    <div>
      <button
        class="pure-button pure-button-primary"
        data-testid="join-submit"
      >
        Join the game
      </button>
    </div>
  </form>
</div>
`;

exports[`<Unit Test> Join Route should render the form with a game id input if no id is provided 1`] = `
<div
  class="container"
  data-testid="join-page"
>
  <h1>
    Join Game
  </h1>
  <span>
    Enter your info and we can get this party started
  </span>
  <form
    data-testid="join-form"
  >
    <div
      class="stack-y-2"
    >
      <label>
        <span
          class="stack-x-1"
        >
          Game id:
        </span>
        <input
          data-testid="game-id"
          type="text"
        />
      </label>
    </div>
    <div
      class="stack-y-2"
    >
      <label>
        <span
          class="stack-x-1"
        >
          Player name:
        </span>
        <input
          data-testid="player-name"
          type="text"
        />
      </label>
    </div>
    <div>
      <button
        class="pure-button pure-button-primary"
        data-testid="join-submit"
      >
        Join the game
      </button>
    </div>
  </form>
</div>
`;
